frontend/src/reducers/quizReducer.ts

```diff
@@ -0,0 +1,86 @@
+import { Action, InitialState } from "../contexts";
+import { initialState } from "../contexts/initialState";
+import { getUpdatedScore, getUpdatedOptions } from "../utils";
+
+import {
+  INCREMENT_QUESTION_NUMBER,
+  INITIALISE_QUIZ_ATTEMPT,
+  RESET_QUIZ,
+  SELECT_OPTION,
+  SHOW_RESULT,
+  UPDATE_SCORE,
+} from "./actions";
+
+export default function quizReducer(
+  prevState: InitialState,
+  action: Action
+): InitialState {
+  switch (action.type) {
+    case INITIALISE_QUIZ_ATTEMPT:
+      return {
+        ...prevState,
+        quizAttempt: action.payload.quiz,
+      };
+
+    case SELECT_OPTION:
+      if (prevState.quizAttempt) {
+        const { questionId, optionId } = action.payload;
+        return {
+          ...prevState,
+          quizAttempt: {
+            ...prevState.quizAttempt,
+            questions: getUpdatedOptions(
+              prevState.quizAttempt.questions,
+              questionId,
+              optionId
+            ),
+          },
+        };
+      }
+      throw new Error("No quiz found!");
+
+    case INCREMENT_QUESTION_NUMBER:
+      if (prevState.quizAttempt) {
+        return {
+          ...prevState,
+          currentQuestionNumber: prevState.currentQuestionNumber + 1,
+        };
+      }
+      throw new Error("No quiz found!");
+
+    case UPDATE_SCORE:
+      if (prevState.quizAttempt) {
+        const { questionId } = action.payload;
+        const prevScore = prevState.quizAttempt.score;
+        const score = getUpdatedScore(prevState.quizAttempt, questionId);
+        return {
+          ...prevState,
+          quizAttempt: {
+            ...prevState.quizAttempt,
+            score: score,
+          },
+          correct:
+            score > prevScore
+              ? prevState.correct + 1
+              : score < prevScore
+              ? prevState.correct - 1
+              : prevState.correct,
+        };
+      }
+      throw new Error("No quiz found!");
+
+    case SHOW_RESULT:
+      return {
+        ...prevState,
+        showReview: true,
+      };
+
+    case RESET_QUIZ:
+      return {
+        ...initialState,
+      };
+
+    default:
+      throw new Error("Invalid Action");
```

Comment:
In the reducer function above you have defined the return type as InitialState. Here you are throwing an error, instead of throwing an error it is better to return the state i.e., prevState.

Reply:
Actually, Tanay had told this in one of the office hours code reviews that you should let the user go into an unhandled state like the `default` one & suggested to throw error to handle such actions

Reply:
Ok, then it is fine.